lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js

```diff
@@ -181,6 +183,284 @@
         }
     });
 
+    platform.registerChakraLibraryFunction("DefaultMergeSort", function(array, length) {
```

Comment:
This method handles the no comparator function case. I made it a whole separate function to make the logic simpler.

As the default comparison is to compare strings I make an array of Strings of every element in the input array first - as converting for each comparison was slower and more costly.

Reply:
The specification says to do a ToString() coercion on the two items to be compared for every comparison.  Can ToString() have side effects?  If so your optimization will be an observable spec violation.

Reply:
ToString() can have side effects, but I believe that that's listed as a situation where the resulting data structure is implementation-defined.

Reply:
@Penguinwizzard What I meant was, the specification says the default compare function should do ToString() on each operand each time it’s called, while @rhuanjl claims to have ToString()’d the whole array in advance to compare the stringified items directly, and I wasn’t sure if that was a spec violation or not.

It’s probably fine in practice since the actual order and number of comparisons is already implementation-defined.

Reply:
The sorting algorithm is implementation defined so no one can rely on the number of toString calls - and every element would always need toString'ing at least once.

Reply:
If the spec allows it, doing this computation ahead of time seems like a good optimization.

Reply:
Pretty sure it's fine per spec BUT - it's still not fast enough I think I need to make my mechanism for default compare better before you can merge this - whilst the results with a provided compare method are comparable (per the numbers in the gist above) currently there is a large regression with the default compare.

Reply:
The default compare "intentionally" (by design) causes numbers to sort as strings -- obviously a custom comparison function will achieve better performance. What do you have in mind to improve the default comparator? Seems like the ToString operation would be the bottleneck.

Reply:
Running it through a profiler at the moment the bottleneck actually seems to be the less than operation being used on the strings I'm not sure if either there isn't a Jit fast path for it OR it isn't being hit- I got similar performance scrapping the cacheing and implementing a method in JsBuiltinEngineInterfaceObject.cpp to convert to string and compare - if I have a method that just compares combined with cacheing I may be a little better - if that doesn't work I have a few other ideas.

Reply:
Yeah, the `<` operator doesn't seem to have any fast path for strings. I tried a small JS file that used `<` on strings a bunch and it generated code that:

1. Checks whether both operands are ints (they never were, but I guess maybe it pays to be optimistic sometimes)
2. If not, branches to a helper block
3. In the helper block, calls Op_Less, which will check whether the operands are numbers again and then switch based on their type
4. Checks whether there was an implicit call and bails out if so

I don't think this speed regression should block merging this PR, because I expect most callers pass a comparator function anyway, and stable output is a very nice upgrade. I could imagine two options for improvement:

* Expose a native method for string-string comparisons and call it here
* Change the lowerer to generate nicer code for comparisons between likely-string values (which could benefit code outside this method too)

Code for `<` operation:

```
 GLOBOPT INSTR:                        BrLt_A         $L9, s28[LikelyCanBeTaggedValue_String].var, s32[LikelyCanBeTaggedValue_String].var #0000   Func # (#1.1), #2 Bailout: #0009 Func  (#1.1), #2 (BailOutOnImplicitCalls)


 1E4D781120F   s107(r15).i64 = MOV    s28(r10)[LikelyCanBeTaggedValue_String].var
 1E4D7811212   s107(r15).i64 = SHR    s107(r15).i64, 48 (0x30).i8
 1E4D7811216   s108(rbx).i64 = MOV    s32(r11)[LikelyCanBeTaggedValue_String].var
 1E4D7811219   s108(rbx).i64 = SHR    s108(rbx).i64, 32 (0x20).i8
 1E4D781121D   s107(r15).i64 = OR     s107(r15).i64, s108(rbx).i64
 1E4D7811220           CMP            s107(r15).i32, 65537 (0x10001).i32
 1E4D7811227           JNE            $L24
 1E4D781122D           CMP            s28(r10)[LikelyCanBeTaggedValue_String].i32, s32(r11)[LikelyCanBeTaggedValue_String].i32
 1E4D7811230           JLT            $L9
 1E4D7811232  $L25:
 1E4D7811232   s162(rbx).i32 = XOR    s162(rbx).i32, s162(rbx).i32             Func # (#1.1), #2
```

The helper at L24 looks like this:

```
 1E4D7811584  $L24: [helper]
 1E4D7811584   s28<-32>.var = MOV     (r10).var                                Func # (#1.1), #2
 1E4D7811588   s32<-40>.var = MOV     (r11).var                                Func # (#1.1), #2
 1E4D781158C   s168<-80>.i64 = MOV    (r8).i64                                 Func # (#1.1), #2
 1E4D7811590   s167<-72>.i64 = MOV    (r9).i64                                 Func # (#1.1), #2
 1E4D7811594   s166<-64>.i64 = MOV    (rdx).i64                                Func # (#1.1), #2
 1E4D7811598   s165<-56>.i64 = MOV    (rcx).i64                                Func # (#1.1), #2
 1E4D781159C   s164<-48>.i64 = MOV    (rax).i64                                Func # (#1.1), #2
 1E4D78115A0   s109(rbx).var = MOV    s32(r11)[LikelyCanBeTaggedValue_String].var  Func # (#1.1), #2
 1E4D78115A3   s110(r15).var = MOV    s28(r10)[LikelyCanBeTaggedValue_String].var  Func # (#1.1), #2
 1E4D78115A6   arg3(s113)(r8).u64 = MOV  0x01DCD5DCC300 (ScriptContext).u64
 1E4D78115B0   arg2(s114)(rdx).var = MOV  s109(rbx).var
 1E4D78115B3   arg1(s115)(rcx).var = MOV  s110(r15).var
 1E4D78115B6   s116(rax).u64 = MOV    Op_Less.u64
 1E4D78115C0   s99(rbx).u64 = MOV     s99<-24>.u64
 1E4D78115C4   [s99(rbx).u64 <0x01DCD5DC1FE0 (&ImplicitCallFlags)>].u8 = MOV  1 (0x1).i8
 1E4D78115C7   s28<-32>.var = MOV     s28(r10).var
 1E4D78115CB           NOP            3 (0x3).i8
 1E4D78115CE   s32<-40>.var = MOV     s32(r11).var
 1E4D78115D2   s112(rax).i64 = CALL   s116(rax).u64                            Func # (#1.1), #2
 1E4D78115D5   s111(rbx).i64 = MOV    s112(rax).i64
 1E4D78115D8   s99(r11).u64 = MOV     s99<-24>.u64                             Func # (#1.1), #2
 1E4D78115DC           CMP            [s99(r11).u64 <0x01DCD5DC1FE0 (&ImplicitCallFlags)>].u8, 1 (0x1).i8  Func # (#1.1), #2
 1E4D78115E0   (rax).i64 = MOV        s164<-48>.i64
 1E4D78115E4   (rcx).i64 = MOV        s165<-56>.i64
 1E4D78115E8   (rdx).i64 = MOV        s166<-64>.i64
 1E4D78115EC   (r9).i64 = MOV         s167<-72>.i64
 1E4D78115F0   (r8).i64 = MOV         s168<-80>.i64
 1E4D78115F4   (r11).var = MOV        s32<-40>.var
 1E4D78115F8   (r10).var = MOV        s28<-32>.var
 1E4D78115FC           JEQ            $L26                                     Func # (#1.1), #2
 1E4D78115FE  $L27: [helper]
 1E4D78115FE   s28<-32>.var = MOV     s28(r10).var
 1E4D7811602   s32<-40>.var = MOV     s32(r11).var
 1E4D7811606   (rdx).i64 = MOV        s111(rbx).i64                            Func # (#1.1), #2
 1E4D7811609   (rcx).u64 = MOV        0x01E4D75EDC60 (BailOutRecord).u64       Func # (#1.1), #2
 1E4D7811613   (rax).u64 = MOV        SaveAllRegistersAndBranchBailOut.u64     Func # (#1.1), #2
 1E4D781161D           CALL           (rax).u64                                Func # (#1.1), #2 Bailout: #0009 Func  (#1.1), #2 (BailOutOnImplicitCalls)
 1E4D7811620           JMP            $L23
 1E4D7811625  $L26: [helper]
 1E4D7811625           TEST           s111(rbx).i64, s111(rbx).i64
 1E4D7811628           JNE            $L9                                      Func # (#1.1), #2
 1E4D781162E           JMP            $L25
```

Reply:
We might also want to add the capability to know that the array is entirely strings; then we'd be able to do this particular operation significantly more quickly, and it wouldn't surprise me if there's a good number of rwc situations where arrays of densely packed strings are common.